### PR TITLE
feat: add disableWindowAnimation caps

### PR DESCRIPTION
**Pull request by pr4bh4sh**
Address https://github.com/appium/appium-espresso-driver/issues/534 and makes animation configurable before launching the AUT.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -179,6 +179,7 @@ appium:chromeOptions | A mapping, that allows to customize chromedriver options.
 Capability Name | Description
 --- | ---
 appium:disableSuppressAccessibilityService | Being set to `true` tells the instrumentation process to not suppress accessibility services during the automated test. This might be useful if your automated test needs these services. `false` by default
+appium:setAnimation | Sets the animation for Android device under test. `false` by default
```
Comment (KazuCocoa) [inline README.md:182]: Could we use `disableWindowAnimation` by default `true` to keep it consistent with UIA2 as well? (if this animation had no issue)

File: lib/driver.js [hunk 2]
```diff
@@ -393,8 +393,19 @@ class EspressoDriver extends BaseDriver {
     // start an avd, set the language/locale, pick an emulator, etc...
     // TODO with multiple devices we'll need to parameterize this
     await helpers.initDevice(this.adb, this.opts);
+    this.caps.setAnimation = this.caps.setAnimation || false;
     // https://github.com/appium/appium-espresso-driver/issues/72
-    if (await this.adb.isAnimationOn()) {
+    if (this.caps.setAnimation && await this.adb.isAnimationOn()) {
```
Comment (mykola-mokhnach) [inline lib/driver.js:398]: consider only calling `await this.adb.isAnimationOn()`. Each extra adb call takes time.

File: lib/driver.js [hunk 3]
```diff
@@ -393,8 +393,19 @@ class EspressoDriver extends BaseDriver {
...
+      this.log.debug('Animation is already on');
+      this.wasAnimationEnabled = true;
+    } else if (this.caps.setAnimation && (await this.adb.isAnimationOn() === false)) {
+      try {
+        await this.adb.setAnimationState(true);
+        this.wasAnimationEnabled = false;
+      } catch (err) {
+        this.log.warn(`Unable to turn on animations: ${err.message}`);
+      }
+    } else if (this.caps.setAnimation === false && await this.adb.isAnimationOn()) {
```
Comment (mykola-mokhnach) [inline lib/driver.js:408]: I don't see a need for an explicit `false` equality check after `this.caps.setAnimation = this.caps.setAnimation || false` has been done above. 

Also, is it really necessary to set it explicitly to false?

File: lib/driver.js [hunk 4]
```diff
@@ -393,8 +393,19 @@ class EspressoDriver extends BaseDriver {
...
-    if (await this.adb.isAnimationOn()) {
+    if (this.caps.setAnimation && await this.adb.isAnimationOn()) {
+      this.log.debug('Animation is already on');
+      this.wasAnimationEnabled = true;
+    } else if (this.caps.setAnimation && (await this.adb.isAnimationOn() === false)) {
+      try {
+        await this.adb.setAnimationState(true);
+        this.wasAnimationEnabled = false;
+      } catch (err) {
+        this.log.warn(`Unable to turn on animations: ${err.message}`);
```
Comment (mykola-mokhnach) [inline lib/driver.js:406]: we should not silence this error if the animation enablement has been explicitly asked in capabilities.

File: lib/driver.js [hunk 5]
```diff
@@ -393,8 +393,19 @@ class EspressoDriver extends BaseDriver {
     // start an avd, set the language/locale, pick an emulator, etc...
     // TODO with multiple devices we'll need to parameterize this
     await helpers.initDevice(this.adb, this.opts);
+    this.caps.setAnimation = this.caps.setAnimation || false;
     // https://github.com/appium/appium-espresso-driver/issues/72
-    if (await this.adb.isAnimationOn()) {
+    if (this.caps.setAnimation && await this.adb.isAnimationOn()) {
+      this.log.debug('Animation is already on');
+      this.wasAnimationEnabled = true;
```
Comment (mykola-mokhnach) [inline lib/driver.js:400]: if animation was already on then we should not change its state afterwards

File: lib/driver.js [hunk 6]
```diff
@@ -393,8 +393,19 @@ class EspressoDriver extends BaseDriver {
     // start an avd, set the language/locale, pick an emulator, etc...
     // TODO with multiple devices we'll need to parameterize this
     await helpers.initDevice(this.adb, this.opts);
+    this.caps.setAnimation = this.caps.setAnimation || false;
```
Comment (mykola-mokhnach) [inline lib/driver.js:396]: maybe it makes sence to extract this logic into a separate method as it's getting a but more lengthy and complicated

File: lib/driver.js (hunk 3 above)
Comment (pr4bh4sh) [reply lib/driver.js]: refactor, the state is true unless `disableWindowAnimation` is explicitly set to false.

File: lib/driver.js (hunk 4 above)
Comment (pr4bh4sh) [reply lib/driver.js]: removed the try catch

File: lib/driver.js (hunk 6 above)
Comment (pr4bh4sh) [reply lib/driver.js]: extracted into a function

File: lib/driver.js [hunk 7]
```diff
@@ -393,15 +393,11 @@ class EspressoDriver extends BaseDriver {
     // start an avd, set the language/locale, pick an emulator, etc...
     // TODO with multiple devices we'll need to parameterize this
     await helpers.initDevice(this.adb, this.opts);
+
+    // default state is window animation disabled
+    this.caps.disableWindowAnimation = this.caps.disableWindowAnimation !== false;
```
Comment (mykola-mokhnach) [inline lib/driver.js:398]: I don't see a need to override the cap value. It is enough to call `await this.setWindowAnimationState(this.caps.disableWindowAnimation !== false);`

File: lib/driver.js [hunk 8]
```diff
@@ -463,6 +459,24 @@ class EspressoDriver extends BaseDriver {
     await this.addDeviceInfoToCaps();
   }
 
+  async setWindowAnimationState(disableAnimation) {
```
Comment (mykola-mokhnach) [inline lib/driver.js:462]: the argument name seems confusing to me. I would rather change it to `isEnabled`

Comment (pr4bh4sh) [reply lib/driver.js:462]: the default state for the variable is `true`.
`disableAnimation === true` animation is disabled
`disableAnimation === false` animation is enabled

`isEnabled === true` sounds like animation should be enabled but when the cap `disableWindowAnimation` value is true animation should be disabled
and wise versa for the `false` value of `disableWindowAnimation`.

Should I just change it to `animationState`. We are anyway setting the value true by `this.caps.disableWindowAnimation !== false` except for the explicit `false` value.

Comment (mykola-mokhnach) [reply lib/driver.js]: it does not matter here the cap name is. It is about the arg name for this particular method. Eventually, `state` assumes multiple possible states, but this one could be on or off (true/false)

Comment (mykola-mokhnach) [reply lib/driver.js]: > isEnabled === true sounds like animation should be enabled but when the cap disableWindowAnimation value is true animation should be disabled

Right, simply pass `disableWindowAnimation === false` to it (inverted condition)

Comment (pr4bh4sh) [reply lib/driver.js]: Done

File: lib/driver.js [hunk 9]
```diff
@@ -463,6 +458,24 @@ class EspressoDriver extends BaseDriver {
     await this.addDeviceInfoToCaps();
   }
 
+  async setWindowAnimationState(isEnabled) {
+    const isAnimationOn = await this.adb.isAnimationOn();
+    const shouldDisableAnimation = !isEnabled && isAnimationOn;
+    const shouldEnableAnimation = isEnabled && !isAnimationOn;
+
+    if (shouldDisableAnimation) {
+      this.log.debug('Disabling window animation as "disableWindowAnimation" capability is set to true');
```
Comment (mykola-mokhnach) [inline lib/driver.js:467]: it might not be necessarily set to true (e.g. if the default value is used)

**Review by mykola-mokhnach — APPROVED** [review]

File: README.md [hunk 10]
```diff
@@ -179,6 +179,7 @@ appium:chromeOptions | A mapping, that allows to customize chromedriver options.
 Capability Name | Description
 --- | ---
 appium:disableSuppressAccessibilityService | Being set to `true` tells the instrumentation process to not suppress accessibility services during the automated test. This might be useful if your automated test needs these services. `false` by default
+appium:disableWindowAnimation | Disables the window animation for Android device under test. `true` by default
```
Comment (mykola-mokhnach) [inline README.md:182]: Perhaps it makes sense to describe in more details how enabling or disabling animations affects Espresso tests

**Review by KazuCocoa — APPROVED** [review]

File: lib/driver.js (hunk 9 above)
Comment (pr4bh4sh) [reply lib/driver.js]: updated the message

File: README.md (hunk 10 above)
Comment (pr4bh4sh) [reply README.md]: Added link for google espresso testing recommendation https://developer.android.com/training/testing/espresso/setup#set-up-environment

**Pull request merged**